File: apps/yawp/test/yawp/canonical_json_test.exs
defmodule Yawp.CanonicalJsonTest do
          use ExUnit.Case, async: true

  alias Yawp.CanonicalJson

  @fixture_path Path.expand(
                  "../../priv/test_vectors/canonical_json.json",
                  __DIR__
                )

  @vectors @fixture_path |> File.read!() |> Jason.decode!() |> Map.fetch!("vectors")

  for v <- @vectors do
    @vector v
    test "vector: " <> @vector["name"] do
      assert CanonicalJson.encode(@vector["input"]) == @vector["output"]
    end
  end

  test "object key order is independent of insertion order" do
    assert CanonicalJson.encode(%{"b" => 1, "a" => 2}) ==
             CanonicalJson.encode(%{"a" => 2, "b" => 1})
  end

  test "round-trip stability: encode(decode(encode(x))) == encode(x)" do
    x = %{"z" => [1, %{"b" => 2, "a" => "hi"}], "a" => nil}
    once = CanonicalJson.encode(x)
    {:ok, parsed} = CanonicalJson.decode(once)
    twice = CanonicalJson.encode(parsed)
    assert once == twice
  end

  test "atom keys are encoded the same as string keys" do
    assert CanonicalJson.encode(%{a: 1, b: 2}) ==
             CanonicalJson.encode(%{"a" => 1, "b" => 2})
  end
end

File: apps/yawp/assets/app/__tests__/canonical-json.test.ts
/**
 * Shared canonical-JSON (RFC 8785) test vectors. The exact same fixture is
 * consumed by the Elixir test suite (test/yawp/canonical_json_test.exs) —
 * byte-for-byte agreement between platforms is the invariant.
 */

import {canonicalJson} from '../canonical-json';
import fixture from '../../../priv/test_vectors/canonical_json.json';

interface Vector {
  name: string;
  input: unknown;
  output: string;
}

describe('canonicalJson (RFC 8785)', () => {
  for (const v of fixture.vectors as Vector[]) {
    test(v.name, () => {
      expect(canonicalJson(v.input)).toBe(v.output);
    });
  }

  test('object key order is independent of insertion order', () => {
    expect(canonicalJson({b: 1, a: 2})).toBe(canonicalJson({a: 2, b: 1}));
  });

  test('round-trip stability: encode(JSON.parse(encode(x))) === encode(x)', () => {
    const x = {z: [1, {b: 2, a: 'hi'}], a: null};
    const once = canonicalJson(x);
    const twice = canonicalJson(JSON.parse(once));
    expect(twice).toBe(once);
  });

  test('throws on non-finite numbers', () => {
    expect(() => canonicalJson(Infinity)).toThrow();
    expect(() => canonicalJson(NaN)).toThrow();
  });
});
